remotes/docker/fetcher.go

```diff
@@ -150,6 +151,28 @@ func (r dockerFetcher) Fetch(ctx context.Context, desc ocispec.Descriptor) (io.R
 	})
 }
 
+func httpError(req *request, resp *http.Response) error {
+	var registryErr Errors
+
+	buf := &bytes.Buffer{}
+	_, err := io.Copy(buf, resp.Body)
+	if err == nil {
+		err = json.NewDecoder(buf).Decode(&registryErr)
+	}
+
+	if err != nil || registryErr.Len() < 1 {
+		// If decoding fails or registryErr looks unhelpful, return the original payload as is.
+		s := buf.String()
+		// But only if the payload looks helpful.
+		if len(s) > 0 {
+			s = ": " + s
+		}
+		return fmt.Errorf("unexpected status code %v: %v%s", req.String(), resp.Status, s)
```

Comment:
Wondering if it could make sense to return the response as-is in the "not valid JSON" case, wdyt?

(agreed on the limit reader though, to prevent _really_ bad responses)

Reply:
Without `unexpected status code ...`? I'd like to keep the URL (from req.String) and HTTP status code. The response body may have these, but it isn't guaranteed.

Reply:
Yeah. Tricky one indeed. I guess either makes sense ("we couldn't parse; we don't know so here's what the registry told us" vs adding the extra information). I guess all bets are off in either case, so 🤷‍♂️.